File: Navigation/nav2_option/g1_sdk_bridge/tests/test_protocol.py
import unittest

import _pathfix  # noqa: F401
from protocol import MAGIC, BridgeStatus, CmdPacket, StatePacket, make_cmd


class TestCmdPacket(unittest.TestCase):
    def test_roundtrip(self):
        pkt = make_cmd(seq=42, vx=0.2, vy=0.0, omega=-0.1)
        decoded = CmdPacket.decode(pkt.encode())
        self.assertEqual(decoded, pkt)

    def test_decode_rejects_wrong_size(self):
        with self.assertRaises(ValueError):
            CmdPacket.decode(b"\x00" * 4)

    def test_decode_rejects_bad_magic(self):
        pkt = make_cmd(seq=1, vx=0.0, vy=0.0, omega=0.0)
        raw = bytearray(pkt.encode())
        raw[0:4] = (MAGIC ^ 0xFFFFFFFF).to_bytes(4, "little")
        with self.assertRaises(ValueError):
            CmdPacket.decode(bytes(raw))

    def test_age_seconds_is_nonnegative_and_grows(self):
        pkt = make_cmd(seq=1, vx=0.0, vy=0.0, omega=0.0)
        now = pkt.timestamp_ns
        self.assertEqual(pkt.age_seconds(now), 0.0)
        self.assertAlmostEqual(pkt.age_seconds(now + 100_000_000), 0.1, places=6)

    def test_age_seconds_clamped_to_zero_if_future(self):
        # 時刻の巻き戻り(単調時計なので理論上起きないはずだが)に対しても負にならないことを保証する
        pkt = make_cmd(seq=1, vx=0.0, vy=0.0, omega=0.0)
        self.assertEqual(pkt.age_seconds(pkt.timestamp_ns - 1_000_000), 0.0)


class TestStatePacket(unittest.TestCase):
    def test_roundtrip(self):
        pkt = StatePacket(
            seq=7,
            timestamp_ns=123456789,
            status=BridgeStatus.NAVIGATING,
            x=1.5,
            y=-2.5,
            yaw=0.78,
            vx=0.1,
            vy=0.0,
            omega=0.05,
            sdk_error_count=2,
        )
        decoded = StatePacket.decode(pkt.encode())
        self.assertEqual(decoded, pkt)
        self.assertIsInstance(decoded.status, BridgeStatus)

    def test_decode_rejects_wrong_size(self):
        with self.assertRaises(ValueError):
            StatePacket.decode(b"\x00" * 10)


if __name__ == "__main__":
    unittest.main()

File: Navigation/nav2_option/g1_sdk_bridge_cpp/tests/test_protocol.cpp
#include "g1_sdk_bridge/protocol.hpp"

#include <gtest/gtest.h>

#include <cstring>
#include <vector>

using namespace g1_sdk_bridge;

TEST(CmdPacket, Roundtrip) {
    const CmdPacket pkt = MakeCmd(42, 0.2, 0.0, -0.1);
    const CmdWire wire = pkt.Encode();
    const CmdPacket decoded = CmdPacket::Decode(&wire, sizeof(wire));
    EXPECT_TRUE(decoded == pkt);
}

TEST(CmdPacket, DecodeRejectsWrongSize) {
    std::vector<std::uint8_t> raw(4, 0);
    EXPECT_THROW(CmdPacket::Decode(raw.data(), raw.size()), ProtocolError);
}

TEST(CmdPacket, DecodeRejectsBadMagic) {
    const CmdPacket pkt = MakeCmd(1, 0.0, 0.0, 0.0);
    CmdWire wire = pkt.Encode();
    wire.magic ^= 0xFFFFFFFFu;
    EXPECT_THROW(CmdPacket::Decode(&wire, sizeof(wire)), ProtocolError);
}

TEST(CmdPacket, AgeSecondsIsNonNegativeAndGrows) {
    const CmdPacket pkt = MakeCmd(1, 0.0, 0.0, 0.0);
    const std::uint64_t now = pkt.timestamp_ns;
    EXPECT_DOUBLE_EQ(pkt.AgeSeconds(now), 0.0);
    EXPECT_NEAR(pkt.AgeSeconds(now + 100'000'000ULL), 0.1, 1e-6);
}

TEST(CmdPacket, AgeSecondsClampedToZeroIfFuture) {
    // 時刻の巻き戻り(単調時計なので理論上起きないはずだが)に対しても負にならないことを保証する
    const CmdPacket pkt = MakeCmd(1, 0.0, 0.0, 0.0);
    EXPECT_DOUBLE_EQ(pkt.AgeSeconds(pkt.timestamp_ns - 1'000'000ULL), 0.0);
}

TEST(StatePacket, Roundtrip) {
    StatePacket pkt;
    pkt.seq = 7;
    pkt.timestamp_ns = 123456789;
    pkt.status = BridgeStatus::kNavigating;
    pkt.x = 1.5;
    pkt.y = -2.5;
    pkt.yaw = 0.78;
    pkt.vx = 0.1;
    pkt.vy = 0.0;
    pkt.omega = 0.05;
    pkt.sdk_error_count = 2;

    const StateWire wire = pkt.Encode();
    const StatePacket decoded = StatePacket::Decode(&wire, sizeof(wire));
    EXPECT_TRUE(decoded == pkt);
}

TEST(StatePacket, DecodeRejectsWrongSize) {
    std::vector<std::uint8_t> raw(10, 0);
    EXPECT_THROW(StatePacket::Decode(raw.data(), raw.size()), ProtocolError);
}
